src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,231 @@
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys which have
+    /// sufficient budget to consume. For those which do not pass, match keys are discarded and not passed to the output.
+    pub fn get_eligible_packets_with_budget<'a>(
+        &'a mut self,
+        packets: impl Iterator<Item = &'a BlindedMatchKey>,
+        budget: u32,
+    ) -> impl Iterator<Item = &'a BlindedMatchKey> {
+        packets.filter(
+            move |packet| match self.increase_budget_consumed(packet, budget) {
+                Ok(true) => true,
+                Ok(false) | Err(_) => false,
+            },
+        )
+    }
+
+    pub fn get_total_privacy_budget(&self) -> u32 {
+        self.total_privacy_budget
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
+        let total_privacy_budget = self.get_total_privacy_budget();
+
+        Ok(transaction(
+            &mut self.redis_connection,
+            &[packet],
+            |con, pipe| {
+                let val: u32 = con.get(&packet).unwrap_or(0);
+                if val + budget > total_privacy_budget {
+                    pipe.set(&packet, val).ignore().query::<()>(con)?;
+                    Ok(Some(false))
+                } else {
+                    // increment
+                    pipe.set(&packet, val + budget).ignore().query::<()>(con)?;
+                    Ok(Some(true))
+                }
+            },
+        )?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use lazy_static::lazy_static;
+    use rand::thread_rng;
+    use std::sync::Mutex;
+
+    lazy_static! {
+        static ref PRIVACY_BUDGET: Mutex<PrivacyBudget> =
+            Mutex::new(PrivacyBudget::new("127.0.0.1", "", false).unwrap());
+    }
+
+    impl Drop for BlindedMatchKey {
+        fn drop(&mut self) {
+            clear_redis_key(self);
+        }
+    }
+
+    fn clear_redis_key(packet: &BlindedMatchKey) {
+        let mut privacy_budget = PRIVACY_BUDGET.lock().unwrap();
+        privacy_budget
+            .redis_connection
+            .del::<&BlindedMatchKey, ()>(packet)
+            .expect("Failed to delete keys");
+    }
+
+    fn create_ciphertexts() -> [BlindedMatchKey; 4] {
+        let mut rng = thread_rng();
+
+        let k1 = DecryptionKey::new(&mut rng);
+        let mk1 = RistrettoPoint::random(&mut rng);
+        let mk2 = RistrettoPoint::random(&mut rng);
+        let mk3 = RistrettoPoint::random(&mut rng);
+        let mk4 = RistrettoPoint::random(&mut rng);
+        let r = Scalar::random(&mut rng);
+
+        let c1 = BlindedMatchKey::new(k1, mk1, r);
+        let c2 = BlindedMatchKey::new(k1, mk2, r);
+        let c3 = BlindedMatchKey::new(k1, mk3, r);
+        let c4 = BlindedMatchKey::new(k1, mk4, r);
```

Comment:
Yeah, this is not how I think about Blinded Match Keys. I do not think they are CipherText, they are not encrypted.